include/xsparse/level_capabilities/co_iteration.hpp

```diff
@@ -0,0 +1,200 @@
+                                            ? std::get<0>(*std::get<I>(t1))
+                                            : min_ik... });
+                }
+
+            public:
+                using iterator_category = std::forward_iterator_tag;
+                using reference = typename std::
+                    tuple<IK, std::tuple<std::optional<typename Levels::BaseTraits::PK>...>>;
+
+                explicit inline iterator(
+                    coiteration_helper const& coiterHelper,
+                    std::tuple<typename Levels::LevelCapabilities::iteration_helper::iterator...>
+                        it) noexcept
+                    : m_coiterHelper(coiterHelper)
+                    , iterators(it)
+                {
+                    min_helper(iterators,
+                               std::apply([&](auto&... args) { return std::tuple(args.end()...); },
+                                          m_coiterHelper.m_iterHelpers));
+                }
+
+                inline auto get_PKs() const noexcept
+                {
+                    return std::apply([&](auto&... args)
+                                      { return std::make_tuple(locate(args)...); },
+                                      this->iterators);
+                }
+
+                template <class iter>
+                inline auto locate(iter i) const noexcept
+                {
+                    return (std::get<0>(*i) == min_ik)
+                               ? std::optional<std::tuple_element_t<1, decltype(*i)>>(
+                                   std::get<1>(*i))
+                               : std::nullopt;
+                }
+
+                template <class iter>
+                inline void advance_iter(iter& i) const noexcept
+                {
+                    if (static_cast<IK>(std::get<0>(*i)) == min_ik)
+                    {
+                        ++i;
+                    }
+                }
+
+                inline reference operator*() const noexcept
+                {
+                    auto PK_tuple = get_PKs();
+                    return { min_ik, PK_tuple };
+                }
+
+                inline iterator operator++(int) const noexcept
+                {
+                    iterator tmp = *this;
+                    ++(*this);
+                    return tmp;
+                }
+
+                inline iterator& operator++() noexcept
+                {
+                    std::apply([&](auto&... args) { ((advance_iter(args)), ...); }, iterators);
+                    min_helper(iterators,
+                               std::apply([&](auto&... args) { return std::tuple(args.end()...); },
+                                          m_coiterHelper.m_iterHelpers));
+                    return *this;
+                }
+
+                inline bool operator!=(iterator const& other) const noexcept
+                {
+                    return !m_coiterHelper.m_coiterate.m_comparisonHelper(
```

Comment:
Can this perhaps be replaced by a fold expression?

Reply:
Fold expressions for the arguments of the function?

Reply:
If it's not easily done, let's ignore it for now.

Reply:
Okay!